### PR TITLE
Add an initial particle flow implementation

### PR DESCRIPTION
**Pull request by therwig**
This PR aims to implement a skeleton of the necessary requirements to perform particle reconstruction across multiple sub-systems.  Currently, this considers (truth) tracks, and clusters in each of the Ecal and Hcal subsystems.

The PR is related to updates to the ECal and HCal submodules in order to utilize a common CaloCluster base-class:
https://github.com/LDMX-Software/Ecal/pull/53
https://github.com/LDMX-Software/Hcal/pull/65

Performance was shown in past eN and weekly meetings including an overview talk by Jason Wu (a summer intern at FNAL in '23) presented at the end of his program here:
https://docs.google.com/presentation/d/18yqZ_82kGT8PvL1rOExaX4xFmR0oYgylV1ESbV1pXSE/edit?usp=sharing

## Submodule PRs
- https://github.com/LDMX-Software/Hcal/pull/65
- https://github.com/LDMX-Software/Ecal/pull/53

### CODE REVIEW

File: Recon/src/Recon/DBScanClusterBuilder.cxx
Comment (tvami) [inline Recon/src/Recon/DBScanClusterBuilder.cxx]:
```suggestion
      ldmx_log(debug) << "trying " << i;
```

File: Recon/src/Recon/DBScanClusterBuilder.cxx
Comment (tvami) [inline Recon/src/Recon/DBScanClusterBuilder.cxx]:
```suggestion
	 ldmx_log(debug) << "- starting a cluster from " << i;
```

File: Recon/src/Recon/DBScanClusterBuilder.cxx
Comment (tvami) [inline Recon/src/Recon/DBScanClusterBuilder.cxx]:
```suggestion
	    ldmx_log(debug) << "== tried " << j;
```

File: Recon/src/Recon/DBScanClusterBuilder.cxx
Comment (tvami) [inline Recon/src/Recon/DBScanClusterBuilder.cxx]:
```suggestion
	     ldmx_log(debug) << "== used " << j;
```

File: Recon/src/Recon/DBScanClusterBuilder.cxx
Comment (tvami) [inline Recon/src/Recon/DBScanClusterBuilder.cxx]:
```suggestion
   ldmx_log(debug)
```
(and no need for the ` << endl`)

File: Recon/include/Recon/DBScanClusterBuilder.h
Comment (tvami) [inline Recon/include/Recon/DBScanClusterBuilder.h]:
```suggestion
  /// Enable logging
  enableLogging("DBScanClusterBuilder")
```

File: Recon/include/Recon/DBScanClusterBuilder.h
Comment (tvami) [inline Recon/include/Recon/DBScanClusterBuilder.h]:
```suggestion
#include "Framework/EventProcessor.h"
```

File: Recon/src/Recon/Event/PFCandidate.cxx
Comment (tvami) [inline Recon/src/Recon/Event/PFCandidate.cxx]:
```suggestion
              << "Ecal energy: " << ecalEnergy_ << " MeV, "
```

File: Recon/src/Recon/Event/PFCandidate.cxx
Comment (tvami) [inline Recon/src/Recon/Event/PFCandidate.cxx]:
```suggestion
              << "Hcal energy: " << hcalEnergy_ << " MeV, "
```

File: Recon/src/Recon/Event/PFCandidate.cxx
Comment (tvami) [inline Recon/src/Recon/Event/PFCandidate.cxx]: is there any reason why the `Print()` starts with uppercase letter? all the other functions start with lowercase, it would be nice to have consistency in the style. @tomeichlersmith ?

File: Recon/src/Recon/PFEcalClusterProducer.cxx
Comment (tvami) [inline Recon/src/Recon/PFEcalClusterProducer.cxx]: with the modifications proposed above, I think these should not be needed anymore

File: Recon/src/Recon/PFHcalClusterProducer.cxx
Comment (tvami) [inline Recon/src/Recon/PFHcalClusterProducer.cxx]: Same as above

File: Recon/src/Recon/PFTrackProducer.cxx
Comment (tvami) [inline Recon/src/Recon/PFTrackProducer.cxx]: should we shoot a msg to the user that this failed? "Input track collection not found" or something?

File: Recon/src/Recon/ParticleFlow.cxx
Comment (tvami) [inline Recon/src/Recon/ParticleFlow.cxx:22]: in general when an object is `new`-ed, it should also have a `delete` somewhere in the end of the code to avoid memory leak. I didnt didnt find the delete, but maybe you do it somewhere else?

File: Recon/test/particleSources.py
Comment (tvami) [inline Recon/test/particleSources.py]: should this not come from some existing library's definition of pi? To avoid some comparison differences when somebody will do this with one more digit of pi and have slightly different result?

File: Recon/include/Recon/ParticleFlow.h
Comment (tvami) [inline Recon/include/Recon/ParticleFlow.h]: again for consistency in the naming, these should be lowercase

File: Recon/src/Recon/PFTruthProducer.cxx
Comment (tvami) [inline Recon/src/Recon/PFTruthProducer.cxx]:
```suggestion
```

**Review by tvami — CHANGES_REQUESTED** [review]
Thanks a lot @therwig for this development! I have a few comments about logging msg, one comment about memory management and a few style comments. None of this is a big deal but I think it's better to catch them now than to run into issues in the far future. Thanks!

**Comment by therwig** [comment]
Thanks Tamas, I believe that I'm happy with all of your suggestions -- will implement them and ping back here.

File: Recon/src/Recon/PFTrackProducer.cxx
Comment (therwig) [reply Recon/src/Recon/PFTrackProducer.cxx]: done!

File: Recon/src/Recon/ParticleFlow.cxx
Comment (therwig) [reply Recon/src/Recon/ParticleFlow.cxx:22]: indeed I missed this (thanks!) its now added to the process end

File: Recon/test/particleSources.py
Comment (therwig) [reply Recon/test/particleSources.py]: now I do `from math import pi`

File: Recon/include/Recon/ParticleFlow.h
Comment (therwig) [reply Recon/include/Recon/ParticleFlow.h]: done

**Comment by therwig** [comment]
OK @tvami , I believe I've made all of the changes that you've requested now. I checked that things run with a clean compile from source and generated a cocktail of 1000 events (cocktail of random e/pi/mu/p/n), seeing no failures.

File: Recon/src/Recon/Event/PFCandidate.cxx
Comment (therwig) [reply Recon/src/Recon/Event/PFCandidate.cxx]: fixed now

**Review by tvami — APPROVED** [review]
Perfect, thanks Christian!

**Comment by tomeichlersmith** [comment]
The Ecal changes have been merged and included in v0.2.8, please do
```
cd Ecal
git fetch --tags
git checkout v0.2.8
cd ..
git add Ecal
git commit -m "Ecal on v0.2.8"
```

**Comment by tvami** [comment]
There seems to be a merge conflict now:

Screenshot 2024-04-17 at 11 39 17

**Comment by therwig** [comment]
Hmm, I see it too but I'm not sure what the conflict _actually is_. I was just following Tom's instructions! 😬

**Comment by tomeichlersmith** [comment]
(This is a _rebase_ conflict and not a _merge_ conflict. I like doing rebase since it keeps the commit history clean and linear, but it tends to generate more conflicts since it is attempting to re-apply **all** commits from your history back onto `trunk`. In general, this is solved by `git rebase trunk` after pulling the latest trunk, but then you'd have to resolve the conflicts. Testing locally, I can `git merge trunk` without any conflicts.)

I am willing to drop my personal `rebase` preference in this case since it is a new addition anyways. I can manually do a `git merge` locally instead of having GitHub attempt a `git rebase` merge on their servers.

**Comment by tomeichlersmith** [comment]
For anyone interested, locally, I am doing what GitHub does on their servers after the merge button has been pressed.
```
# make sure trunk is up to date
git switch trunk
git pull origin trunk
# make sure dev branch is up to date
git switch pfReco
git pull origin pfReco
git merge trunk
# merge dev branch into trunk
git switch trunk
git merge pfReco
# push to GitHub
git push origin trunk
```
To add some extra cool-ness, I _do not_ need to click a button saying this PR was merged. GitHub detects that the branch was merged and will mark this PR merged for me. :sunglasses:

**Pull request merged**
